### ISSUE TITLE
Anisotropy results do not agree with BoneJ1

### ISSUE THREAD

**Issue by mdoube** [bonej-org/BoneJ2#182]
**Describe the bug**
DA and vectors from new Anisotropy plugin do not agree with results from BoneJ1

**To Reproduce**
1. Open [emu.tif](http://bonej.org/images/emu.tif)
2. Run Anisotropy in ImageJ1 / BoneJ1
3. Run Anisotropy in ImageJ2

Note that BoneJ2 consistently calculates DA lower (~0.68) than BoneJ1 (~0.8), although BoneJ2's result is more internally consistent from run to run.

BoneJ1's results are validated by displaying the best-fit ellipsoid alongside the image data, and by rotating the image data to be parallel with the fabric tensor. It is also possible that there is a bug in BoneJ1's implementation.

**Expected behavior**
DA and eigenvalues/vectors  should be the same or similar for the same image.

**Comment by mdoube** [comment]
Validated against two test images, designed to give high and low DA:
Low DA (≃ 0.0) [binary_noise.tif](https://github.com/bonej-org/BoneJ2/files/3151330/binary_noise.tif.zip)
Hgh DA (≃ 0.9), created with this macro
```java
newImage("Untitled", "8-bit black", 512, 512, 512);
setColor(255,255,255);

//planes in xy, every 32 slices
for (i = 1; i <= nSlices; i+=32){
	setSlice(i);
	fill();
}

//planes in yz every 8 columns
for (i = 7; i < getWidth(); i+=8){
	for (z = 1; z <= nSlices; z++){
	   setSlice(z);
	   drawLine(i, 0, i, getHeight()-1);
	}
}

//planes in xz every 16 rows
for (i = 15; i < getHeight(); i+=16){
	for (z = 1; z <= nSlices; z++){
	   setSlice(z);
	   drawLine(0, i, getWidth()-1, i);
	}
}
```
Both legacy and modern give low and high DA values where they should, so agree, but their eigenvalues / radii do not agree. Eigenvectors agree for the plates test image, but with each orthogonal unit vector in different sorting. Vectors do not agree for the noise test image, perhaps because orientation should have a strong random effect in the isotropic case.

**Comment by rimadoma** [comment]
I don't know if they can agree. The methods are so different, e. g. in mine some vectors are always going to miss foreground

**Comment by mdoube** [comment]
Conceptually they are pretty similar: run a bunch of vectors with the same orientation but different translations through the image and count boundaries and sampling lengths. Maybe there are effects relating to sampling? I tuned mine to e.g. not be affected by the rectangular pixel grid (which it is if your sampling distance is too short IIRC)

**Comment by mdoube** [comment]
I'd like to do the same visualisation of the rose plot, fitted ellipsoid and rotation to fabric tensor for a visual check.

**Comment by alessandrofelder** [comment]
Does the difference in calculating DA (`1-ev_min/ev_max` versus `1-sqrt(ev_min/ev_max)`) between the BoneJ1 and BoneJ2 versions explain the disagreement to any extent?

**Comment by mdoube** [comment]
> Does the difference in calculating DA (`1-ev_min/ev_max` versus `1-sqrt(ev_min/ev_max)`) between the BoneJ1 and BoneJ2 versions explain the disagreement to any extent?

Yes, to some extent. But there is still disagreement in mid-range values, e.g. for [this image of trabeculae from a dog calcaneus](https://github.com/bonej-org/BoneJ2/files/3154963/dog_calcaneus_trabeculae_binary.tif.zip) BoneJ1 DA ≃ 0.68 and BoneJ2 DA ≃ 0.53 even after using eigenvalue ratios.

**Comment by mdoube** [comment]
> some vectors are always going to miss foreground
That's true - I remember being stuck with what to do if a vector made it to length = sphere radius without hitting a boundary, and IIRC the solution was to count it as one (or something) just to avoid having a divide-by-zero and NaN situation. But ideally you'd add up the full length of the sampling vectors in a direction and count all the boundary hits in that direction, and if there are still zero either excluding the point or setting the MIL for that direction to the sum of lengths.

**Comment by mdoube** [comment]
I think this is all down to the pixel sample spacing: if it's too small you start to get interference from the pixel grid. Will take a closer look and probably just tweak the wrapper a little bit. Will share my analysis after lunch / presently.

**Comment by alessandrofelder** [comment]
Taking the risk of complicating things unnecessarily: Does [this](https://github.com/bonej-org/BoneJ2/blob/7e2a8c71a302616fe31ad62dd73d2d9f8c426e47/Modern/ops/src/main/java/org/bonej/ops/mil/MILPlane.java#L177) explain the difference to any extent?
```
	private static <B extends BooleanType<B>> long countPhaseChanges(
		final RandomAccessible<B> interval, final Vector3d start,
		final Vector3dc gap, final long samples)

	{
		final RandomAccess<B> access = interval.randomAccess();
		boolean previous = false;
		long phaseChanges = 0;
		for (long i = 0; i < samples; i++) {
			final boolean current = getVoxel(access, start);
			if (current && !previous) {
				phaseChanges++;
			}
			previous = current;
			start.add(gap);
		}
		return phaseChanges;
	}
```
It counts only changes from foreground to background (and not those from background to foreground).
This definition follows the MIL definition of [ Moreno et al, 2012](http://www.diva-portal.org/smash/get/diva2%3A544192/FULLTEXT01.pdf) (See Fig. 1).

**Comment by mdoube** [comment]
>  Does [this](https://github.com/bonej-org/BoneJ2/blob/7e2a8c71a302616fe31ad62dd73d2d9f8c426e47/Modern/ops/src/main/java/org/bonej/ops/mil/MILPlane.java#L177) explain the difference to any extent?

Yes potentially. BoneJ1 does it like this:
```java
for (double pos = -radius; pos <= radius; pos += vectorSampling) {
       // find the index of the voxel that the sample falls
       // within offset from centroid
       final int x = (int) Math.round(pos * vXvW);
       final int y = (int) Math.round(pos * vYvH);
       final int z = (int) Math.round(pos * vZvD);
       final int testIndex = centroidIndex + b * z + a * y + x;
       // determine if the voxel is thresholded or not
       if (workArray[testIndex] == 0) {
        thisPos = true;
       } else {
        thisPos = false;
       }
       // if this pos is not equal to last pos then an
       // interface is counted
       if (thisPos != lastPos) {
        nIntercepts++;
       }
       // then before incrementing the for loop, set
       // lastPos to thisPos
       lastPos = thisPos;
}
```
I prefer this approach because anisotropy of both phases can contribute to the final result. Imagine a situation where trabeculae oriented in one direction were thicker than trabeculae in another orientation, but that the distance between medial axes was the same. If you measure only one half of the phase changes the anisotropy measurement will miss the extra information about anisotropy in one of the phases.

**Comment by mdoube** [comment]
It seems like radii measured with BoneJ2 are ≃ 2× BoneJ1 radii, even after fixing the issue in the last comment.

**Comment by mdoube** [comment]
Closed by #187
